drivers/hmis/spec/models/hmis/form/definition_spec.rb

```diff
@@ -69,4 +69,31 @@
       expect(Hmis::Form::Definition.find_definition_for_role(fd1.role, project: p1, version: 1)).to eq(fd2)
     end
   end
+
+  describe 'with funder and project type instances' do
+    let(:role) { :SERVICE }
+    it 'applies correct specificity (project > org > funder&ptype > funder > ptype)' do
+      base_fd = Hmis::Form::Definition.find_definition_for_role(role) # created by hmis base setup
+
+      p1 = create(:hmis_hud_project, project_type: 1)
+      p2 = create(:hmis_hud_project, project_type: 1, funders: [43])
+      p3 = create(:hmis_hud_project, project_type: 2, funders: [43])
+      p4 = create(:hmis_hud_project, project_type: 2)
+      p5 = create(:hmis_hud_project, project_type: 1, funders: [43])
+      p6 = create(:hmis_hud_project, project_type: 1, funders: [43])
+
+      fi1 = create(:hmis_form_instance, role: role, entity: nil, project_type: 1, funder: nil)
+      fi2 = create(:hmis_form_instance, role: role, entity: nil, project_type: 1, funder: 43)
+      fi3 = create(:hmis_form_instance, role: role, entity: nil, project_type: nil, funder: 43)
+      fi4 = create(:hmis_form_instance, role: role, entity: p5)
+      fi5 = create(:hmis_form_instance, role: role, entity: p6.organization)
+
+      expect(Hmis::Form::Definition.find_definition_for_role(role, project: p1)).to eq(fi1.definition)
+      expect(Hmis::Form::Definition.find_definition_for_role(role, project: p2)).to eq(fi2.definition)
+      expect(Hmis::Form::Definition.find_definition_for_role(role, project: p3)).to eq(fi3.definition)
+      expect(Hmis::Form::Definition.find_definition_for_role(role, project: p4)).to eq(base_fd)
+      expect(Hmis::Form::Definition.find_definition_for_role(role, project: p5)).to eq(fi4.definition)
+      expect(Hmis::Form::Definition.find_definition_for_role(role, project: p6)).to eq(fi5.definition)
+    end
```

Comment:
I added another commit https://github.com/greenriver/hmis-warehouse/pull/3346/commits/9ee20a567029fe85f364a643aa801283945a7126 that splits out the scope `for_project_by_type_or_funder` into 3 scopes: `for_project_by_project_type`, `for_project_by_funder`, and `for_project_by_funder_and_project_type`. It has the benefit of making the scopes more readable, but mainly this lets us prioritize more specific instances (project type AND funder) over more general instances (just project type, or just funder).

Reply:
We don't have a use case like that yet, but I'm sure it will come up. For example "all emergency shelters should collect X service, but ones funded by Y should use this different form that collects additional info needed by the funder" or whatnot